framework/include/problems/SubProblem.h

```diff
@@ -1129,25 +1230,94 @@ SubProblem::addFunctor(const std::string & name,
                        const Moose::FunctorBase<T> & functor,
                        const THREAD_ID tid)
 {
+  constexpr bool added_functor_is_ad =
+      !std::is_same<T, typename MetaPhysicL::RawType<T>::value_type>::value;
+
   mooseAssert(tid < _functors.size(), "Too large a thread ID");
 
+  auto & functor_to_request_info = _functor_to_request_info[tid];
   auto & functors = _functors[tid];
   auto it = functors.find("wraps_" + name);
   if (it != functors.end())
   {
     // We have this functor already. If it's a null functor, we want to replace it with the valid
     // functor we have now. If it's not then we'll add a new entry into the multimap and then we'll
-    // error later if a user requests a functor because their request is ambiguous
-    auto * const existing_wrapper = dynamic_cast<Moose::Functor<T> *>(it->second.get());
+    // error later if a user requests a functor because their request is ambiguous. This is the
+    // reason that the functors container is a multimap: for nice error messages
+    auto * const existing_wrapper_base =
+        added_functor_is_ad ? std::get<2>(it->second).get() : std::get<1>(it->second).get();
+    auto * const existing_wrapper = dynamic_cast<Moose::Functor<T> *>(existing_wrapper_base);
     if (existing_wrapper && existing_wrapper->template wrapsType<Moose::NullFunctor<T>>())
     {
+      // Sanity check
+      auto [request_info_it, request_info_end_it] = functor_to_request_info.equal_range(name);
+      if (request_info_it == request_info_end_it)
+        mooseError("We are wrapping a NullFunctor but we don't have any unfilled functor request "
+                   "info. This doesn't make sense.");
+
+      // Check for valid requests
+      while (request_info_it != request_info_end_it)
+      {
+        auto & [requested_functor_is_ad, requestor_is_ad] = request_info_it->second;
+        if (!requested_functor_is_ad && requestor_is_ad && added_functor_is_ad)
+          mooseError("We are requesting a non-AD functor from an AD object, but the true functor "
+                     "is AD. This means we could be dropping important derivatives. We will not "
+                     "allow this");
+        request_info_it = functor_to_request_info.erase(request_info_it);
```

Comment:
we're deleting the request because we've filled it and we dont care anymore?
maybe add a comment?